File: Sources/CatDesktopPet/PetStateMachine.swift
import Foundation

enum PetState: String {
    case hidden
    case idle
    case sleeping
    case clicked
    case catchingBug
    case idleAction
    case socialInteraction
    case grabbed
    case dropped
}

enum PetEvent {
    case show
    case hide
    case click
    case mouseOverPet
    case idleActionDue
    case nearbyPet
    case reactionFinished
    case dragStarted
    case dragEnded
    case sleep
    case wake
}

final class PetStateMachine {
    private(set) var state: PetState = .hidden
    var onStateChanged: ((PetState) -> Void)?

    private var resetToken = UUID()

    func send(_ event: PetEvent) {
        switch event {
        case .show:
            transition(to: .idle)
        case .hide:
            transition(to: .hidden)
        case .sleep:
            guard state == .idle else {
                return
            }

            transition(to: .sleeping)
        case .wake:
            guard state == .sleeping else {
                return
            }

            transition(to: .idle)
        case .click:
            guard state != .grabbed, state != .hidden else {
                return
            }

            transition(to: .clicked)
        case .mouseOverPet:
            guard state == .idle else {
                return
            }

            transition(to: .catchingBug)
        case .idleActionDue:
            guard state == .idle else {
                return
            }

            transition(to: .idleAction)
        case .nearbyPet:
            guard state == .idle else {
                return
            }

            transition(to: .socialInteraction)
        case .reactionFinished:
            guard state == .clicked || state == .catchingBug || state == .idleAction || state == .socialInteraction else {
                return
            }

            transition(to: .idle)
        case .dragStarted:
            guard state != .hidden else {
                return
            }

            transition(to: .grabbed)
        case .dragEnded:
            guard state == .grabbed else {
                return
            }

            transition(to: .dropped)
            returnToIdle(after: 0.24)
        }
    }

    private func transition(to nextState: PetState) {
        guard state != nextState else {
            return
        }

        state = nextState
        resetToken = UUID()
        onStateChanged?(nextState)
    }

    private func returnToIdle(after delay: TimeInterval) {
        let token = resetToken

        DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self] in
            guard let self, self.resetToken == token, self.state != .hidden else {
                return
            }

            self.transition(to: .idle)
        }
    }
}

File: windows/src/shared/pet-state-machine.ts
export type PetState =
  | "hidden"
  | "idle"
  | "sleeping"
  | "clicked"
  | "catchingBug"
  | "idleAction"
  | "socialInteraction"
  | "grabbed"
  | "dropped";

export type PetEvent =
  | "show"
  | "hide"
  | "click"
  | "mouseOverPet"
  | "idleActionDue"
  | "nearbyPet"
  | "reactionFinished"
  | "reactionUnavailable"
  | "dragStarted"
  | "dragEnded"
  | "sleep"
  | "wake";

export type StateChanged = (state: PetState) => void;
export type ReturnToIdleScheduler = (callback: () => void, delayMs: number) => void;

export class PetStateMachine {
  #state: PetState = "hidden";
  #token = 0;
  readonly #onStateChanged?: StateChanged;
  readonly #scheduleReturnToIdle: ReturnToIdleScheduler;

  constructor(
    onStateChanged?: StateChanged,
    scheduleReturnToIdle: ReturnToIdleScheduler = (callback, delayMs) => {
      setTimeout(callback, delayMs);
    }
  ) {
    this.#onStateChanged = onStateChanged;
    this.#scheduleReturnToIdle = scheduleReturnToIdle;
  }

  get state() {
    return this.#state;
  }

  send(event: PetEvent) {
    switch (event) {
      case "show":
        this.#transitionTo("idle");
        break;
      case "hide":
        this.#transitionTo("hidden");
        break;
      case "sleep":
        if (this.#state === "idle") {
          this.#transitionTo("sleeping");
        }
        break;
      case "wake":
        if (this.#state === "sleeping") {
          this.#transitionTo("idle");
        }
        break;
      case "click":
        if (this.#state !== "grabbed" && this.#state !== "hidden") {
          this.#transitionTo("clicked");
        }
        break;
      case "mouseOverPet":
        if (this.#state === "idle") {
          this.#transitionTo("catchingBug");
        }
        break;
      case "idleActionDue":
        if (this.#state === "idle") {
          this.#transitionTo("idleAction");
        }
        break;
      case "nearbyPet":
        if (this.#state === "idle") {
          this.#transitionTo("socialInteraction");
        }
        break;
      case "reactionFinished":
        if (["sleeping", "clicked", "catchingBug", "idleAction", "socialInteraction"].includes(this.#state)) {
          this.#transitionTo("idle");
        }
        break;
      case "reactionUnavailable":
        if (["clicked", "catchingBug", "idleAction", "socialInteraction"].includes(this.#state)) {
          this.#returnToIdle(120);
        }
        break;
      case "dragStarted":
        if (this.#state !== "hidden") {
          this.#transitionTo("grabbed");
        }
        break;
      case "dragEnded":
        if (this.#state === "grabbed") {
          this.#transitionTo("dropped");
          this.#returnToIdle(240);
        }
        break;
    }
  }

  #transitionTo(nextState: PetState) {
    if (this.#state === nextState) {
      return;
    }

    this.#state = nextState;
    this.#token += 1;
    this.#onStateChanged?.(nextState);
  }

  #returnToIdle(delayMs: number) {
    const token = this.#token;

    this.#scheduleReturnToIdle(() => {
      if (this.#token === token && this.#state !== "hidden") {
        this.#transitionTo("idle");
      }
    }, delayMs);
  }
}
